### PR TITLE
build: copy symlinks as symlinks for sti 

### PR DESCRIPTION
**Pull request by wozniakjan**
This PR is a follow up for:
https://github.com/openshift/source-to-image/pull/910
https://github.com/openshift/source-to-image/issues/892
https://github.com/openshift/origin/issues/19809

The `make update-deps` bumped automatically:
- github.com/openshift/api
- k8s.io/client-go
- k8s.io/kubernetes
- github.com/openshift/source-to-image

And s2i brings along a few more changes, not sure if this is desired or if we should rather backport https://github.com/openshift/source-to-image/pull/910 to 3.10?

fixes https://github.com/openshift/origin/issues/19809

### CODE REVIEW

**Comment by bparees** [comment]
> And s2i brings along a few more changes, not sure if this is desired or if we should rather backport 

bringing in the other changes is fine...mostly it was refactoring, a little bit of it is the dockerfile generation logic which is irrelevant to openshift for now.

@wozniakjan can you add a test for this so we don't regress the symlink behavior in the future?

**Comment by bparees** [comment]
Also @wozniakjan @adambkaplan can you think of any scenario where this behavior change would break an existing user?  My previous logic was that this would only break someone who has a git repo that contains a symlink that points outside their git repo (previously that would result in us copying in the external file, now it will likely result in a broken symlink after copying, or a symlink that points to the file in their image's filesystem rather than a copy of the file that existed in our s2i-builder image), and therefore it's unlikely to break anyone and if it does, they were doing something they shouldn't have been doing..but i would definitely appreciate a sanity check.

**Comment by wozniakjan** [comment]
> @wozniakjan can you add a test for this so we don't regress the symlink behavior in the future?

my naive way of testing this was to run 
```bash
$ ln -s [some_file] link
$ oc start-build bc --from-dir=.
$ docker run -it [build_output_image] /bin/bash -c "if [ -L link ]; then echo 'ok'; fi"
```
Is it reasonable to transform this into an extended test and instead of `docker run` use `oc run` if possible?

**Comment by bparees** [comment]
> Is it reasonable to transform this into an extended test and instead of docker run use oc run if possible?

yeah that sounds fine.

File: pkg/build/builder/sti.go [hunk 1]
```diff
@@ -223,6 +223,8 @@ func (s *S2IBuilder) Build() error {
 		CGroupLimits:              s.cgLimits,
 		ScriptDownloadProxyConfig: scriptDownloadProxyConfig,
 		BlockOnBuild:              true,
+
+		KeepSymlinks: true,
```
Comment (adambkaplan) [inline pkg/build/builder/sti.go]: needs `go fmt`

**Comment by adambkaplan** [comment]
@bparees I agree that if someone was relying on the symlink behavior to copy data from the s2i build container to the resulting image, that's beyond the bounds of supported behavior.

**Comment by wozniakjan** [comment]
@adambkaplan @bparees , one test keeps failing after bumping source-to-image.
https://github.com/openshift/origin/blob/a6fbcb30801497f5406515e0d7ed82d823116105/test/extended/builds/s2i_env.go#L51

I tried some debugging but I am not very familiar with recent changes in s2i repo, what I found out is
it fails with non-existent env variable
https://github.com/openshift/origin/blob/9581804c8c5d4bf557be30f6faabb6f10de48649/test/extended/testdata/builds/sti-environment-build-app/config.ru#L1

If I hardcode a value there, Webrick no longer panics and pod joins the service. Also, source-to-image release-3.10 works well.

**Comment by adambkaplan** [comment]
@bparees @wozniakjan I actually don't know how this test _wasn't_ failing before. The source for this test uses the old `.sti` path to load directories. This was deprecated 3 years ago according to Github blame.

**Comment by bparees** [comment]
> @bparees @wozniakjan The source for this test uses the old .sti path to load directories. This was deprecated 3 years ago, but Ben finally killed it with the Dockerfile build feature.

the fact that we broke ourselves by removing deprecated behavior doesn't bode well for the implications on our users :/

**Comment by adambkaplan** [comment]
Moving to #20490

**Pull request closed without merging**